### ISSUE TITLE
Database Cleaner starting too soon

### ISSUE THREAD

**Issue by bpeicher** [DatabaseCleaner/database_cleaner#330]
I'm having an issue with truncation happening in the middle of an integration test. It deletes test records, and then test continues on until an ActiveRecord:NotFound occurs. Excerpt from log pasted below.

```
[1m[35m (0.1ms)[0m  BEGIN
  [1m[36mUser Exists (0.2ms)[0m  [1mSELECT  1 AS one FROM `users`  WHERE `users`.`slug` = 'fb4b3e44ddf4' LIMIT 1[0m
  [1m[35mUser Exists (0.1ms)[0m  SELECT  1 AS one FROM `users`  WHERE `users`.`email` = '2tuna@wiper.com' LIMIT 1
  [1m[36mUser Exists (0.1ms)[0m  [1mSELECT  1 AS one FROM `users`  WHERE `users`.`remember_token` = 'tuyFEjBZgrQVJVQOK7pFiQ' LIMIT 1[0m
  [1m[35mSQL (0.2ms)[0m  INSERT INTO `users` (`created_at`, `email`, `first_name`, `last_name`, `password_digest`, `prefs`, `remember_token`, `slug`, `updated_at`) VALUES ('2015-01-15 04:13:26', '2tuna@wiper.com', 'Tuna', 'Wiper 2', '$2a$04$pmPj9mzauNHbiUszUbk4teCFDFKSRHMK6Zmy1MyczfR0uu5hPtNEe', '--- !ruby/hash:ActiveSupport::HashWithIndifferentAccess\ndonation_emails: true\n', 'tuyFEjBZgrQVJVQOK7pFiQ', 'fb4b3e44ddf4', '2015-01-15 04:13:26')
  [1m[36m (0.8ms)[0m  [1mCOMMIT[0m
  [1m[35m (0.1ms)[0m  BEGIN
  [1m[36mInvitation Exists (0.2ms)[0m  [1mSELECT  1 AS one FROM `invitations`  WHERE `invitations`.`slug` = '3c66c8586438' LIMIT 1[0m
  [1m[35mSQL (0.2ms)[0m  INSERT INTO `invitations` (`created_at`, `email`, `slug`, `trip_group_id`, `updated_at`, `user_id`) VALUES ('2015-01-15 04:13:26', '1tuna@wiper.com', '3c66c8586438', 1, '2015-01-15 04:13:26', 2)
  [1m[36m (0.5ms)[0m  [1mCOMMIT[0m
  [1m[35m (0.1ms)[0m  SELECT @@FOREIGN_KEY_CHECKS
  [1m[36m (0.1ms)[0m  [1mSET FOREIGN_KEY_CHECKS = 0[0m
  [1m[35m (6.1ms)[0m  TRUNCATE TABLE `addresses`;
  [1m[36m (6.3ms)[0m  [1mTRUNCATE TABLE `administrations`;[0m
  [1m[35m (7.9ms)[0m  TRUNCATE TABLE `bank_accounts`;
  [1m[36m (5.6ms)[0m  [1mTRUNCATE TABLE `buddies`;[0m
  [1m[35m (8.1ms)[0m  TRUNCATE TABLE `campaigns`;
  [1m[36m (12.5ms)[0m  [1mTRUNCATE TABLE `donations`;[0m
  [1m[35m (15.6ms)[0m  TRUNCATE TABLE `error_logs`;
  [1m[36m (80.1ms)[0m  [1mTRUNCATE TABLE `event_logs`;[0m
  [1m[35m (9.7ms)[0m  TRUNCATE TABLE `grants`;
  [1m[36m (78.7ms)[0m  [1mTRUNCATE TABLE `guides`;[0m
  [1m[35m (59.3ms)[0m  TRUNCATE TABLE `invitations`;
  [1m[36m (160.3ms)[0m  [1mTRUNCATE TABLE `invoices`;[0m
  [1m[35m (7.1ms)[0m  TRUNCATE TABLE `organizations`;
  [1m[36m (6.1ms)[0m  [1mTRUNCATE TABLE `transaction_logs`;[0m
  [1m[35m (7.1ms)[0m  TRUNCATE TABLE `trip_groups`;
  [1m[36m (9.4ms)[0m  [1mTRUNCATE TABLE `trips`;[0m
  [1m[35m (10.8ms)[0m  TRUNCATE TABLE `users`;
  [1m[36m (0.1ms)[0m  [1mSET FOREIGN_KEY_CHECKS = 1[0m
Started GET "/signin" for 127.0.0.1 at 2015-01-14 23:13:27 -0500
Processing by SessionsController#new as HTML
  Rendered layouts/_h1.html.erb (0.1ms)
  Rendered sessions/_session_form.html.erb (1.3ms)
  Rendered layouts/_meta.html.erb (0.1ms)
  Rendered layouts/_shim.html.erb (0.1ms)
  Rendered layouts/_scripts.html.erb (0.2ms)
  Rendered layouts/_brand.html.erb (0.2ms)
  [1m[35mUser Load (0.4ms)[0m  SELECT  `users`.* FROM `users`  WHERE `users`.`remember_token` IS NULL LIMIT 1
  Rendered layouts/_nav.html.erb (1.3ms)
  Rendered layouts/_header.html.erb (2.1ms)
  Rendered layouts/_content.html.erb (0.2ms)
  Rendered layouts/_links.html.erb (0.2ms)
  Rendered layouts/_footer.html.erb (0.3ms)
Completed 200 OK in 7ms (Views: 6.0ms | ActiveRecord: 0.4ms)
Started POST "/sessions" for 127.0.0.1 at 2015-01-14 23:13:27 -0500
Processing by SessionsController#create as HTML
  Parameters: {"utf8"=>"✓", "session"=>{"email"=>"1tuna@wiper.com", "password"=>"[FILTERED]"}, "commit"=>"Let me in!"}
  [1m[36mUser Load (0.4ms)[0m  [1mSELECT  `users`.* FROM `users`  WHERE `users`.`email` = '1tuna@wiper.com' LIMIT 1[0m
  Rendered layouts/_h1.html.erb (0.1ms)
  Rendered sessions/_session_form.html.erb (0.9ms)
  Rendered layouts/_meta.html.erb (0.1ms)
  Rendered layouts/_shim.html.erb (0.0ms)
  Rendered layouts/_scripts.html.erb (0.2ms)
  Rendered layouts/_brand.html.erb (0.2ms)
  [1m[35mUser Load (0.3ms)[0m  SELECT  `users`.* FROM `users`  WHERE `users`.`remember_token` IS NULL LIMIT 1
  Rendered layouts/_nav.html.erb (0.8ms)
  Rendered layouts/_header.html.erb (1.2ms)
  Rendered layouts/_content.html.erb (0.1ms)
  Rendered layouts/_links.html.erb (0.1ms)
  Rendered layouts/_footer.html.erb (0.3ms)
Completed 200 OK in 6ms (Views: 3.9ms | ActiveRecord: 0.7ms)
Started GET "/i/3c66c8586438/new" for 127.0.0.1 at 2015-01-14 23:13:27 -0500
Processing by CampaignsController#new as HTML
  Parameters: {"invitation_id"=>"3c66c8586438"}
  [1m[36mInvitation Load (0.3ms)[0m  [1mSELECT  `invitations`.* FROM `invitations`  WHERE `invitations`.`slug` = '3c66c8586438' LIMIT 1[0m
Completed 404 Not Found in 2ms

ActiveRecord::RecordNotFound (ActiveRecord::RecordNotFound):
  app/controllers/application_controller.rb:54:in `find_invitation_by_invitation_id'
  test/integration/campaign/campaign_new_share_test.rb:13:in `block in <class:IntegrationTest>'
```

Here is how database cleaner is being invoked in minitest:

```ruby
 def setup
    DatabaseCleaner.strategy = :transaction
    DatabaseCleaner.clean_with :truncation
    DatabaseCleaner.start
  end

  def teardown
    DatabaseCleaner.clean
  end
```
Strange - any ideas?

**Comment by foton** [comment]
I have the same issue. Could not find what triggered truncation in cucumber test. After crossing all buy-checkout procedure, something starts truncating before test is finished.

Side effect of experiments is that I found that 'sleep' of any lengtht in controller will start truncating procedure.

**Comment by bpeicher** [comment]
My issue was related to the minitest_bang gem. Reading my test log, I noted that anything running before truncation began was being created using let!(). This occurs immediately when test begins, prior to any before blocks. Therefore, when the pre-test truncation ran in the before block, it was erasing everything created by let!(). I resorted to removing the bang and using let(), manually invoking objects in a before block.

Instead of this:

```ruby
let!(:person) { Factory :person }
```

I did this:

```ruby
let (:person) { Factory :person }

before { person }
```
